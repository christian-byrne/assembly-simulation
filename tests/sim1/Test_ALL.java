/*
 * Author: Joseph Marbella
 * Description: 52 test cases for sub and add
 */

class Test_ALL {
  public static void print_binary(String prefix, int val) {
    int i;

    System.out.printf("%s", prefix);

    for (i = 31; i >= 0; i--) {
      System.out.printf("%d", (val >> i) & 0x1);

      if (i % 4 == 0 && i != 0)
        System.out.printf(" ");
    }

    System.out.println();
  }

  public static int to_int(RussWire[] data) {
    int result = 0;

    for (byte i = 0; i < 32; ++i) {
      if (data[i].get()) {
        result |= (1 << i);
      }
    }

    return result;
  }

  public static void print_data(int a, int b, Sim1_ADD adder, Sim1_SUB sub) {
    int sum = adder != null ? to_int(adder.sum) : to_int(sub.sum);
    System.out.printf("Decimal:\n");
    System.out.printf("    %11d\n", a);
    System.out.printf("  %c %11d\n", adder == null ? '-' : '+', b);
    System.out.printf(" --------------\n");
    System.out.printf("    %11d\n", sum);
    System.out.printf("  aNonNeg=%d, bNonNeg=%d, sumNonNeg=%d\n", a >= 0 ? 1 : 0, b >= 0 ? 1 : 0,
        adder != null ? ((a + b) >= 0 ? 1 : 0) : ((a - b) >= 0 ? 1 : 0));

    /*
     * This is commented out because not everyone's adder in the SUB class is named
     * "add".
     * Thus, testing for overflow and carryout are omitted. If your adder passes the
     * testAdd cases,
     * then we can assume that the carryOut and overflow will pass these testSub
     * cases as well.
     */

    // System.out.printf(" carryOut=%d\n",
    // adder != null ? (adder.carryOut.get() ? 1 : 0) : (sub.add.carryOut.get() ? 1
    // : 0));
    // System.out.printf(" overflow=%d\n",
    // adder != null ? (adder.overflow.get() ? 1 : 0) : (sub.add.overflow.get() ? 1
    // : 0));

    System.out.printf("\n");

    System.out.printf("Binary:\n");
    print_binary("    ", a);
    print_binary("  + ", adder != null ? b : -b);
    System.out.printf(" -------------------------------------------\n");
    print_binary("    ", sum);
    System.out.printf("\n\n --------------------------------------------------------------------------------------\n");
  }

  public static void testAdd(int a, int b) {
    int aCopy = a, bCopy = b;
    final Sim1_ADD adder = new Sim1_ADD();
    for (byte i = 0; i < 32; ++i) {
      adder.a[i].set((a & 1) != 0);
      adder.b[i].set((b & 1) != 0);
      a >>= 1;
      b >>= 1;
    }
    System.out.println();
    adder.execute();
    print_data(aCopy, bCopy, adder, null);
  }

  public static void testSub(int a, int b) {
    int aCopy = a, bCopy = b;
    final Sim1_SUB sub = new Sim1_SUB();
    for (byte i = 0; i < 32; ++i) {
      sub.a[i].set((a & 1) != 0);
      sub.b[i].set((b & 1) != 0);
      a >>= 1;
      b >>= 1;
    }
    System.out.println();
    sub.execute();
    print_data(aCopy, bCopy, null, sub);
  }

  public static void main(String[] args) {
    testAdd(1, -2);
    testAdd(2, -3);
    testAdd(-3, -3);
    testAdd(Integer.MAX_VALUE, 1);
    testAdd(Integer.MAX_VALUE, Integer.MIN_VALUE);
    testAdd(Integer.MIN_VALUE, -1);
    testAdd(0xCFFFFFFF, 0xCFFFFFFF);
    testAdd(0, -1);
    testAdd(-1, -1);
    testAdd(Integer.MIN_VALUE, -Integer.MIN_VALUE);
    testAdd(0, 0);
    testAdd(Integer.MAX_VALUE, 0);
    testAdd(Integer.MAX_VALUE, Integer.MAX_VALUE);
    testAdd(Integer.MIN_VALUE, 500);
    testAdd(Integer.MAX_VALUE, -Integer.MAX_VALUE);
    testAdd(32, 64);
    testAdd(0xC0000000, 0xC0000000);
    testAdd(3333, 2222);
    testAdd(-3333, 2222);
    testAdd(Integer.MIN_VALUE, 0x80000000);
    testAdd(Integer.MAX_VALUE, -Integer.MIN_VALUE);
    testAdd(Integer.MIN_VALUE, Integer.MAX_VALUE);
    testAdd(-32768, -1);
    testAdd(Integer.MAX_VALUE, 0xFFFFFFFE);
    testAdd(Integer.MAX_VALUE, 0xFFFFFFFF);
    testAdd(Integer.MAX_VALUE, 0xFFFFFFFD);
    testAdd(-3, 5);
    testAdd(-69, 69);
    testAdd(0, 10);
    testAdd(Integer.MAX_VALUE, -2);
    testAdd(-3, Integer.MIN_VALUE);
    testSub(Integer.MAX_VALUE, 1);
    testSub(Integer.MAX_VALUE, Integer.MAX_VALUE);
    testSub(Integer.MAX_VALUE, 0);
    testSub(0, Integer.MAX_VALUE);
    testSub(Integer.MAX_VALUE, -2);
    testSub(-3, Integer.MIN_VALUE);
    testSub(Integer.MAX_VALUE, 1);
    testSub(Integer.MAX_VALUE, 0xFFFFFFFE);
    testSub(Integer.MAX_VALUE, 0xFFFFFFFF);
    testSub(Integer.MAX_VALUE, 0xFFFFFFFD);
    testSub(Integer.MIN_VALUE, -1);
    testSub(Integer.MIN_VALUE, Integer.MAX_VALUE);
    testSub(1, 1);
    testSub(0, 0);
    testSub(-5, 5);
    testSub(-1, 4);
    testSub(1, -1999);
    testSub(0, Integer.MIN_VALUE);
    testSub(55555, 55555);
    testSub(55555, -69);
    testSub(-55555, 69);
    testSub(55555, 666666);
  }
}