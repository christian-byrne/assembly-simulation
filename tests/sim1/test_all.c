#include <stdio.h>
#include <limits.h>
#include <stdbool.h>
#include <stdlib.h>
#include "sim1.h"

void printf_binary(char *prefix, int val);
void print_sim1_data(Sim1Data* data, int isSubtraction) {
  printf("Decimal:\n");
	printf("    %11d\n", data->a);
	printf("  %c %11d\n", isSubtraction ? '-' : '+', data->b);
	printf(" --------------\n");
	printf("    %11d\n", data->sum);
	printf("  aNonNeg=%d, bNonNeg=%d, sumNonNeg=%d\n", data->aNonNeg, data->bNonNeg, data->sumNonNeg);
	printf("  carryOut=%d\n", data->carryOut);
	printf("  overflow=%d\n", data->overflow);
	printf("\n");

	printf       ("Binary:\n");
	printf_binary("    ", data->a);
	printf_binary("  + ", ~data->b + 1);
	printf       (" -------------------------------------------\n");
	printf_binary("    ", data->sum);
  printf       ("\n\n --------------------------------------------------------------------------------------\n");
};

void test_add(int a, int b) {
  Sim1Data* data = calloc(sizeof(Sim1Data), 1);
  
  data->a = a;
  data->b = b;
  data->isSubtraction = 0;

  execute_add(data);

  print_sim1_data(data, 0);
}

void test_sub(int a, int b) {
  Sim1Data* data = calloc(sizeof(Sim1Data), 1);
  
  data->a = a;
  data->b = b;
  data->isSubtraction = 1;

  execute_add(data);

  print_sim1_data(data, 1);
}

int main() {
  // add test cases here
  test_add(1, -2);
  test_add(2, -3);
  test_add(-3, -3);
  test_add(INT_MAX, 1);
  test_add(INT_MAX, INT_MIN);
  test_add(INT_MIN, -1);
  test_add(0xCFFFFFFF, 0xCFFFFFFF);
  test_add(0, -1);
  test_add(-1, -1);
  test_add(INT_MIN, -INT_MIN);
  test_add(0, 0);
  test_add(INT_MAX, 0);
  test_add(INT_MAX, INT_MAX);
  test_add(INT_MIN, 500);
  test_add(INT_MAX, -INT_MAX);
  test_add(32, 64);
  test_add(0xC0000000, 0xC0000000);
  test_add(3333, 2222);
  test_add(-3333, 2222); 
  test_add(INT_MIN, 0x80000000);
  test_add(INT_MAX, -INT_MIN);
  test_add(INT_MIN, INT_MAX);
  test_add(-32768, -1);
  test_add(INT_MAX, 0xFFFFFFFE);
  test_add(INT_MAX, 0xFFFFFFFF);
  test_add(INT_MAX, 0xFFFFFFFD);
  test_add(-3, 5);
  test_add(-69, 69);
  test_add(0, 10);
  test_add(INT_MAX, -2);
  test_add(-3, INT_MIN);
  test_add(INT_MAX, 1);
  test_sub(INT_MAX, INT_MAX);
  test_sub(INT_MAX, 0);
  test_sub(0, INT_MAX);
  test_sub(INT_MAX, -2);
  test_sub(-3, INT_MIN);
  test_sub(INT_MAX, 1);
  test_sub(INT_MAX, 0xFFFFFFFE);
  test_sub(INT_MAX, 0xFFFFFFFF);
  test_sub(INT_MAX, 0xFFFFFFFD);
  test_sub(INT_MIN, -1);
  test_sub(INT_MIN, INT_MAX);
  test_sub(1, 1);
  test_sub(0, 0);
  test_sub(-5, 5);
  test_sub(-1, 4);
  test_sub(1, -1999);
  test_sub(0, INT_MIN);
  test_sub(55555, 55555);
  test_sub(55555, -69);
  test_sub(-55555, 69);
  test_sub(55555, 666666);

  return 0;
}

void printf_binary(char *prefix, int val) {
	int i;

	printf("%s", prefix);

	for (i=31; i>=0; i--)
	{
		printf("%d", (val >> i) & 0x1);

		if (i%4 == 0 && i != 0)
			printf(" ");
	}

	printf("\n");
}