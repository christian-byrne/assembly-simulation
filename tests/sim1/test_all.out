Decimal:
              1
  +          -2
 --------------
             -1
  aNonNeg=1, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 0000 0000 0000 0001
  + 0000 0000 0000 0000 0000 0000 0000 0010
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 1111


 --------------------------------------------------------------------------------------
Decimal:
              2
  +          -3
 --------------
             -1
  aNonNeg=1, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 0000 0000 0000 0010
  + 0000 0000 0000 0000 0000 0000 0000 0011
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 1111


 --------------------------------------------------------------------------------------
Decimal:
             -3
  +          -3
 --------------
             -6
  aNonNeg=0, bNonNeg=0, sumNonNeg=0
  carryOut=1
  overflow=0

Binary:
    1111 1111 1111 1111 1111 1111 1111 1101
  + 0000 0000 0000 0000 0000 0000 0000 0011
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 1010


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  +           1
 --------------
    -2147483648
  aNonNeg=1, bNonNeg=1, sumNonNeg=0
  carryOut=0
  overflow=1

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 1111 1111 1111 1111 1111 1111 1111 1111
 -------------------------------------------
    1000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  + -2147483648
 --------------
             -1
  aNonNeg=1, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 1000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 1111


 --------------------------------------------------------------------------------------
Decimal:
    -2147483648
  +          -1
 --------------
     2147483647
  aNonNeg=0, bNonNeg=0, sumNonNeg=1
  carryOut=1
  overflow=1

Binary:
    1000 0000 0000 0000 0000 0000 0000 0000
  + 0000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    0111 1111 1111 1111 1111 1111 1111 1111


 --------------------------------------------------------------------------------------
Decimal:
     -805306369
  +  -805306369
 --------------
    -1610612738
  aNonNeg=0, bNonNeg=0, sumNonNeg=0
  carryOut=1
  overflow=0

Binary:
    1100 1111 1111 1111 1111 1111 1111 1111
  + 0011 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    1001 1111 1111 1111 1111 1111 1111 1110


 --------------------------------------------------------------------------------------
Decimal:
              0
  +          -1
 --------------
             -1
  aNonNeg=1, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 0000 0000 0000 0000
  + 0000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 1111


 --------------------------------------------------------------------------------------
Decimal:
             -1
  +          -1
 --------------
             -2
  aNonNeg=0, bNonNeg=0, sumNonNeg=0
  carryOut=1
  overflow=0

Binary:
    1111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 1110


 --------------------------------------------------------------------------------------
Decimal:
    -2147483648
  + -2147483648
 --------------
              0
  aNonNeg=0, bNonNeg=0, sumNonNeg=1
  carryOut=1
  overflow=1

Binary:
    1000 0000 0000 0000 0000 0000 0000 0000
  + 1000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
              0
  +           0
 --------------
              0
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 0000 0000 0000 0000
  + 0000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  +           0
 --------------
     2147483647
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=0
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    0111 1111 1111 1111 1111 1111 1111 1111


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  +  2147483647
 --------------
             -2
  aNonNeg=1, bNonNeg=1, sumNonNeg=0
  carryOut=0
  overflow=1

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 1000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 1110


 --------------------------------------------------------------------------------------
Decimal:
    -2147483648
  +         500
 --------------
    -2147483148
  aNonNeg=0, bNonNeg=1, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    1000 0000 0000 0000 0000 0000 0000 0000
  + 1111 1111 1111 1111 1111 1110 0000 1100
 -------------------------------------------
    1000 0000 0000 0000 0000 0001 1111 0100


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  + -2147483647
 --------------
              0
  aNonNeg=1, bNonNeg=0, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0111 1111 1111 1111 1111 1111 1111 1111
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
             32
  +          64
 --------------
             96
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 0000 0000 0010 0000
  + 1111 1111 1111 1111 1111 1111 1100 0000
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0110 0000


 --------------------------------------------------------------------------------------
Decimal:
    -1073741824
  + -1073741824
 --------------
    -2147483648
  aNonNeg=0, bNonNeg=0, sumNonNeg=0
  carryOut=1
  overflow=0

Binary:
    1100 0000 0000 0000 0000 0000 0000 0000
  + 0100 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    1000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
           3333
  +        2222
 --------------
           5555
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 0000 1101 0000 0101
  + 1111 1111 1111 1111 1111 0111 0101 0010
 -------------------------------------------
    0000 0000 0000 0000 0001 0101 1011 0011


 --------------------------------------------------------------------------------------
Decimal:
          -3333
  +        2222
 --------------
          -1111
  aNonNeg=0, bNonNeg=1, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    1111 1111 1111 1111 1111 0010 1111 1011
  + 1111 1111 1111 1111 1111 0111 0101 0010
 -------------------------------------------
    1111 1111 1111 1111 1111 1011 1010 1001


 --------------------------------------------------------------------------------------
Decimal:
    -2147483648
  + -2147483648
 --------------
              0
  aNonNeg=0, bNonNeg=0, sumNonNeg=1
  carryOut=1
  overflow=1

Binary:
    1000 0000 0000 0000 0000 0000 0000 0000
  + 1000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  + -2147483648
 --------------
             -1
  aNonNeg=1, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 1000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 1111


 --------------------------------------------------------------------------------------
Decimal:
    -2147483648
  +  2147483647
 --------------
             -1
  aNonNeg=0, bNonNeg=1, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    1000 0000 0000 0000 0000 0000 0000 0000
  + 1000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 1111


 --------------------------------------------------------------------------------------
Decimal:
         -32768
  +          -1
 --------------
         -32769
  aNonNeg=0, bNonNeg=0, sumNonNeg=0
  carryOut=1
  overflow=0

Binary:
    1111 1111 1111 1111 1000 0000 0000 0000
  + 0000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    1111 1111 1111 1111 0111 1111 1111 1111


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  +          -2
 --------------
     2147483645
  aNonNeg=1, bNonNeg=0, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0010
 -------------------------------------------
    0111 1111 1111 1111 1111 1111 1111 1101


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  +          -1
 --------------
     2147483646
  aNonNeg=1, bNonNeg=0, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    0111 1111 1111 1111 1111 1111 1111 1110


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  +          -3
 --------------
     2147483644
  aNonNeg=1, bNonNeg=0, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0011
 -------------------------------------------
    0111 1111 1111 1111 1111 1111 1111 1100


 --------------------------------------------------------------------------------------
Decimal:
             -3
  +           5
 --------------
              2
  aNonNeg=0, bNonNeg=1, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    1111 1111 1111 1111 1111 1111 1111 1101
  + 1111 1111 1111 1111 1111 1111 1111 1011
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0010


 --------------------------------------------------------------------------------------
Decimal:
            -69
  +          69
 --------------
              0
  aNonNeg=0, bNonNeg=1, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    1111 1111 1111 1111 1111 1111 1011 1011
  + 1111 1111 1111 1111 1111 1111 1011 1011
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
              0
  +          10
 --------------
             10
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 0000 0000 0000 0000
  + 1111 1111 1111 1111 1111 1111 1111 0110
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 1010


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  +          -2
 --------------
     2147483645
  aNonNeg=1, bNonNeg=0, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0010
 -------------------------------------------
    0111 1111 1111 1111 1111 1111 1111 1101


 --------------------------------------------------------------------------------------
Decimal:
             -3
  + -2147483648
 --------------
     2147483645
  aNonNeg=0, bNonNeg=0, sumNonNeg=1
  carryOut=1
  overflow=1

Binary:
    1111 1111 1111 1111 1111 1111 1111 1101
  + 1000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    0111 1111 1111 1111 1111 1111 1111 1101


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  +           1
 --------------
    -2147483648
  aNonNeg=1, bNonNeg=1, sumNonNeg=0
  carryOut=0
  overflow=1

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 1111 1111 1111 1111 1111 1111 1111 1111
 -------------------------------------------
    1000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  -  2147483647
 --------------
              0
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 1000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  -           0
 --------------
     2147483647
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    0111 1111 1111 1111 1111 1111 1111 1111


 --------------------------------------------------------------------------------------
Decimal:
              0
  -  2147483647
 --------------
    -2147483647
  aNonNeg=1, bNonNeg=1, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 0000 0000 0000 0000
  + 1000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    1000 0000 0000 0000 0000 0000 0000 0001


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  -          -2
 --------------
    -2147483647
  aNonNeg=1, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=1

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0010
 -------------------------------------------
    1000 0000 0000 0000 0000 0000 0000 0001


 --------------------------------------------------------------------------------------
Decimal:
             -3
  - -2147483648
 --------------
     2147483645
  aNonNeg=0, bNonNeg=0, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    1111 1111 1111 1111 1111 1111 1111 1101
  + 1000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    0111 1111 1111 1111 1111 1111 1111 1101


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  -           1
 --------------
     2147483646
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 1111 1111 1111 1111 1111 1111 1111 1111
 -------------------------------------------
    0111 1111 1111 1111 1111 1111 1111 1110


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  -          -2
 --------------
    -2147483647
  aNonNeg=1, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=1

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0010
 -------------------------------------------
    1000 0000 0000 0000 0000 0000 0000 0001


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  -          -1
 --------------
    -2147483648
  aNonNeg=1, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=1

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    1000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
     2147483647
  -          -3
 --------------
    -2147483646
  aNonNeg=1, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=1

Binary:
    0111 1111 1111 1111 1111 1111 1111 1111
  + 0000 0000 0000 0000 0000 0000 0000 0011
 -------------------------------------------
    1000 0000 0000 0000 0000 0000 0000 0010


 --------------------------------------------------------------------------------------
Decimal:
    -2147483648
  -          -1
 --------------
    -2147483647
  aNonNeg=0, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    1000 0000 0000 0000 0000 0000 0000 0000
  + 0000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    1000 0000 0000 0000 0000 0000 0000 0001


 --------------------------------------------------------------------------------------
Decimal:
    -2147483648
  -  2147483647
 --------------
              1
  aNonNeg=0, bNonNeg=1, sumNonNeg=1
  carryOut=1
  overflow=1

Binary:
    1000 0000 0000 0000 0000 0000 0000 0000
  + 1000 0000 0000 0000 0000 0000 0000 0001
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0001


 --------------------------------------------------------------------------------------
Decimal:
              1
  -           1
 --------------
              0
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0000 0000 0000 0000 0000 0000 0000 0001
  + 1111 1111 1111 1111 1111 1111 1111 1111
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
              0
  -           0
 --------------
              0
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0000 0000 0000 0000 0000 0000 0000 0000
  + 0000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
             -5
  -           5
 --------------
            -10
  aNonNeg=0, bNonNeg=1, sumNonNeg=0
  carryOut=1
  overflow=0

Binary:
    1111 1111 1111 1111 1111 1111 1111 1011
  + 1111 1111 1111 1111 1111 1111 1111 1011
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 0110


 --------------------------------------------------------------------------------------
Decimal:
             -1
  -           4
 --------------
             -5
  aNonNeg=0, bNonNeg=1, sumNonNeg=0
  carryOut=1
  overflow=0

Binary:
    1111 1111 1111 1111 1111 1111 1111 1111
  + 1111 1111 1111 1111 1111 1111 1111 1100
 -------------------------------------------
    1111 1111 1111 1111 1111 1111 1111 1011


 --------------------------------------------------------------------------------------
Decimal:
              1
  -       -1999
 --------------
           2000
  aNonNeg=1, bNonNeg=0, sumNonNeg=1
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 0000 0000 0000 0001
  + 0000 0000 0000 0000 0000 0111 1100 1111
 -------------------------------------------
    0000 0000 0000 0000 0000 0111 1101 0000


 --------------------------------------------------------------------------------------
Decimal:
              0
  - -2147483648
 --------------
    -2147483648
  aNonNeg=1, bNonNeg=0, sumNonNeg=0
  carryOut=0
  overflow=1

Binary:
    0000 0000 0000 0000 0000 0000 0000 0000
  + 1000 0000 0000 0000 0000 0000 0000 0000
 -------------------------------------------
    1000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
          55555
  -       55555
 --------------
              0
  aNonNeg=1, bNonNeg=1, sumNonNeg=1
  carryOut=1
  overflow=0

Binary:
    0000 0000 0000 0000 1101 1001 0000 0011
  + 1111 1111 1111 1111 0010 0110 1111 1101
 -------------------------------------------
    0000 0000 0000 0000 0000 0000 0000 0000


 --------------------------------------------------------------------------------------
Decimal:
          55555
  -         -69
 --------------
          55624
  aNonNeg=1, bNonNeg=0, sumNonNeg=1
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 1101 1001 0000 0011
  + 0000 0000 0000 0000 0000 0000 0100 0101
 -------------------------------------------
    0000 0000 0000 0000 1101 1001 0100 1000


 --------------------------------------------------------------------------------------
Decimal:
         -55555
  -          69
 --------------
         -55624
  aNonNeg=0, bNonNeg=1, sumNonNeg=0
  carryOut=1
  overflow=0

Binary:
    1111 1111 1111 1111 0010 0110 1111 1101
  + 1111 1111 1111 1111 1111 1111 1011 1011
 -------------------------------------------
    1111 1111 1111 1111 0010 0110 1011 1000


 --------------------------------------------------------------------------------------
Decimal:
          55555
  -      666666
 --------------
        -611111
  aNonNeg=1, bNonNeg=1, sumNonNeg=0
  carryOut=0
  overflow=0

Binary:
    0000 0000 0000 0000 1101 1001 0000 0011
  + 1111 1111 1111 0101 1101 0011 1101 0110
 -------------------------------------------
    1111 1111 1111 0110 1010 1100 1101 1001


 --------------------------------------------------------------------------------------
